scipy/optimize/tests/test_bracket.py

```diff
@@ -286,15 +286,387 @@ def f(x):
+        f = self.init_f()
+
+        xmax = None
+        if other_side:
+            xl0, xm0, xr0 = -xr0, -xm0, -xl0
+            xmin, xmax = None, -xmin if xmin is not None else None
+            lower, middle, upper = -upper, -middle, -lower
+
+        kwargs = self.get_kwargs(
+            xl0=xl0, xr0=xr0, xmin=xmin, xmax=xmax, factor=factor, args=args
+        )
+        result = _bracket_minimum(f, xm0, **kwargs)
+
+        # Check that `nfev` and `nit` have the correct relationship
+        assert result.nfev == result.nit + 3
+        # Check that `nfev` reports the correct number of function evaluations.
+        assert result.nfev == f.count
+        # Check that the number of iterations matches the theoretical value.
+        assert result.nit == n
+
+        # Compare reported bracket to theoretical bracket and reported function
+        # values to function evaluated at bracket.
+        bracket = np.asarray([result.xl, result.xm, result.xr])
+        assert_allclose(bracket, (lower, middle, upper))
+        f_bracket = np.asarray([result.fl, result.fm, result.fr])
+        assert_allclose(f_bracket, f(bracket, *args))
+
+        self.assert_valid_bracket(result)
+        assert result.status == 0
+        assert result.success
+
+    def test_flags(self):
+        # Test cases that should produce different status flags; show that all
+        # can be produced simultaneously
+        def f(xs, js):
+            funcs = [lambda x: (x - 1.5)**2,
+                     lambda x: x,
+                     lambda x: x,
+                     lambda x: np.nan]
+            return [funcs[j](x) for x, j in zip(xs, js)]
+
+        args = (np.arange(4, dtype=np.int64),)
+        xl0, xm0, xr0 = np.full(4, -1.0), np.full(4, 0.0), np.full(4, 1.0)
+        result = _bracket_minimum(f, xm0, xl0=xl0, xr0=xr0,
+                                  xmin=[-np.inf, -1.0, -np.inf, -np.inf],
+                                  args=args, maxiter=3)
+
+        reference_flags = np.array([eim._ECONVERGED, _ELIMITS,
+                                    eim._ECONVERR, eim._EVALUEERR])
+        assert_equal(result.status, reference_flags)
+
+    @pytest.mark.parametrize("minimum", (0.622, [0.622, 0.623]))
+    @pytest.mark.parametrize("dtype", (np.float16, np.float32, np.float64))
+    @pytest.mark.parametrize("xmin", [-5, None])
+    @pytest.mark.parametrize("xmax", [5, None])
+    def test_dtypes(self, minimum, xmin, xmax, dtype):
+        xmin = xmin if xmin is None else dtype(xmin)
+        xmax = xmax if xmax is None else dtype(xmax)
+        minimum = dtype(minimum)
+
+        def f(x, minimum):
+            return ((x - minimum)**2).astype(dtype)
+
+        xl0, xm0, xr0 = np.array([-0.01, 0.0, 0.01], dtype=dtype)
+        result = _bracket_minimum(
+            f, xm0, xl0=xl0, xr0=xr0, xmin=xmin, xmax=xmax, args=(minimum, )
+        )
+        assert np.all(result.success)
+        assert result.xl.dtype == result.xm.dtype == result.xr.dtype == dtype
```

Comment:
For good measure? It is _possible_ that even if `f` returns the right `dtype` we don't preserve it.
```suggestion
        assert result.xl.dtype == result.xm.dtype == result.xr.dtype == dtype
        assert result.fl.dtype == result.fm.dtype == result.fr.dtype == dtype
```